pkg/querier/batch/stream.go

```diff
@@ -159,11 +159,49 @@ func (bs *batchStream) merge(batch *chunk.Batch, size int) {
 			populate(batch, rt)
 			batch.Next()
 		} else {
-			if (rt == chunkenc.ValHistogram || rt == chunkenc.ValFloatHistogram) && lt == chunkenc.ValFloat {
-				// Prefer histograms than floats. Take left side if both have histograms.
-				populate(batch, rt)
-			} else {
+			// The two samples have the same timestamp. They may be the same exact sample or different one
+			// (e.g. we ingested an in-order sample and then OOO sample with same timestamp but different value).
+			// We want query results to be deterministic, so we select one of the two side deterministically.
+			var pickLeftSide bool
+
+			switch {
+			// If the sample types are the same, we pick the highest value.
+			case lt == rt:
+				if lt == chunkenc.ValFloat {
+					_, lValue := bs.curr().At()
+					_, rValue := batch.At()
+					pickLeftSide = lValue > rValue
+				} else if lt == chunkenc.ValHistogram {
+					_, lValue := bs.curr().AtHistogram()
+					_, rValue := batch.AtHistogram()
+
+					pickLeftSide = (*histogram.Histogram)(lValue).Sum > (*histogram.Histogram)(rValue).Sum
```

Comment:
Please .Count instead of Sum.  Observations can be negative, so Sum is not monotonic. However if Count is higher, you could argue that the histogram has more observations ~ more valueable.